feat(core): add utility for resolving defer block information to ng global

crisbeto opened the pull request:
Adds the `getDeferBlocks` function to the global `ng` namespace which returns information about all `@defer` blocks inside of a DOM node. This information can be useful either directly in the browser console or to implement future functionality in the dev tools.

Also includes the following supporting changes:

### refactor(core): remove unused field 
Removes the `prefetchTriggers` field from the `TDeferBlockDetails` since we weren't reading or writing to it anywhere.

### refactor(core): track debugging information about deferred triggers 
Adds a field on the `TDeferBlockDetails` where we can track debugging information about the defer block. Also uses it to store text representation of the different triggers which can be shown to the dev tools.

crisbeto committed:
refactor(core): remove unused field

Removes the `prefetchTriggers` field from the `TDeferBlockDetails` since we weren't reading or writing to it anywhere.

crisbeto committed:
refactor(core): track debugging information about deferred triggers

Adds a field on the `TDeferBlockDetails` where we can track debugging information about the defer block. Also uses it to store text representation of the different triggers which can be shown to the dev tools.

AndrewKushnir commented on packages/core/src/render3/util/defer.ts:
```diff
@@ -0,0 +1,210 @@
+    /** Minimum amount of milliseconds that the block should be shown. */
+    minimumTime: number | null;
+
+    /** Amount of time after which the blockk should be shown. */
```
```suggestion
    /** Amount of time after which the block should be shown. */
```

AndrewKushnir commented on packages/core/src/render3/util/defer.ts:
```diff
@@ -0,0 +1,210 @@
+/*!
```
I'm wondering if we should move this code in the `packages/core/src/defer` folder (for ex. `packages/core/src/defer/debug.ts`).

crisbeto replied:
I put it here since this is where we keep all of the `window.ng` code.

AndrewKushnir replied:
Sounds good. We'd probably need to revisit this (for ex. create a separate folder for all debug-related code), but it's outside of the scope of this PR.

AndrewKushnir commented on packages/core/src/render3/util/defer.ts:
```diff
@@ -0,0 +1,210 @@
+      hasErrorBlock: tDetails.errorTmplIndex !== null,
+      loadingBlock: {
+        exists: tDetails.loadingTmplIndex !== null,
+        minimumTime: tDetails.loadingBlockConfig?.[0] ?? null,
+        afterTime: tDetails.loadingBlockConfig?.[1] ?? null,
```
We can use consts from the `packages/core/src/defer/interfaces.ts` here (the `MINIMUM_SLOT` and `LOADING_AFTER_SLOT` ones).

AndrewKushnir commented on packages/core/src/render3/util/defer.ts:
```diff
@@ -0,0 +1,210 @@
+        exists: tDetails.placeholderTmplIndex !== null,
+        minimumTime: tDetails.placeholderBlockConfig?.[0] ?? null,
+      },
+      triggers: tDetails.debug?.triggers ? Array.from(tDetails.debug.triggers) : [],
```
nit: I'm wondering if we should sort triggers (alphabetically?), so that we group similar triggers (so in DevTools we just show them one after another).

crisbeto replied:
Right now they're sorted in the same order they're defined, aside from `when` which gets declared in the update block so it's always last. I figured that might be easier for users to map back to their source code. That being said, I don't feel strongly against sorting them either.

AndrewKushnir replied:
Yeah, I was thinking about the `when` trigger as well - it'd be placed outside of the order, so we can just group triggers by type (regular, prefetch, hydrate), which would happen automatically if we sort them alphabetically.

crisbeto replied:
I've added the sorting.

AndrewKushnir commented on packages/core/src/render3/util/defer.ts:
```diff
@@ -0,0 +1,210 @@
+    tDetails.hydrateTriggers === null ||
+    tDetails.hydrateTriggers.has(DeferBlockTrigger.Never)
+  ) {
+    return 'not-hydratable';
+  }
+  return registry.has(lDetails[SSR_UNIQUE_ID]) ? 'dehydrated' : 'hydrated';
```
I'm a bit unsure about the states: a block might not use incremental hydration (e.g. do not have `hydrate` triggers set up), thus the placeholder would be SSR'ed and later hydrated on the client. I think we can make it more explicit that this is an incremental hydration state by naming a corresponding property `incrementalHydrationState` (vs `hydrationState`) and renaming one of the return values:

```suggestion
    return 'not-configured';
  }
  return registry.has(lDetails[SSR_UNIQUE_ID]) ? 'dehydrated' : 'hydrated';
```

AndrewKushnir commented on packages/core/src/render3/util/defer.ts:
```diff
@@ -0,0 +1,210 @@
+  const nodes: Node[] = [];
+  let currentTNode = lView[TVIEW].firstChild;
+
+  while (currentTNode !== null) {
```
We'll probably need to use the `collectNativeNodes` function here, which accounts for content projection, i18n, etc.

crisbeto replied:
Ah nice, I was thinking that we likely have a function for this but I couldn't remember the name.

AndrewKushnir commented on packages/core/src/render3/util/global_utils.ts:
```diff
@@ -79,6 +80,7 @@ const globalUtilsFunctions = {
   'getDirectives': getDirectives,
   'applyChanges': applyChanges,
   'isSignal': isSignal,
+  'getDeferBlocks': getDeferBlocks,
```
I'd propose adding the `ɵ` for now (to mark it as private), until we get a chance to test it fully in DevTools:
```suggestion
  'ɵgetDeferBlocks': getDeferBlocks,
```

crisbeto reviewed (commented):
Feedback has been addressed.

AndrewKushnir reviewed (approved):
@crisbeto thanks for implementing this API 👍

crisbeto committed:
feat(core): add utility for resolving defer block information to ng global

Adds the `getDeferBlocks` function to the global `ng` namespace which returns information about all `@defer` blocks inside of a DOM node. This information can be useful either directly in the browser console or to implement future functionality in the dev tools.

AndrewKushnir commented:
This PR was merged into the repository by commit 57f3550219f2a57c7c26c9183e48ee66845e0439.

The changes were merged into the following branches: main
